snapshot/storage/metastore.go

```diff
@@ -0,0 +1,79 @@
+package storage
+
+import (
+	"context"
+
+	"github.com/docker/containerd/snapshot"
+)
+
+// MetaStore is used to store metadata related to a snapshot driver. The
+// MetaStore is intended to store metadata related to name, state and
+// parentage. Using the MetaStore is not required to implement a snapshot
+// driver but can be used to handle the persistence and transactional
+// complexities of a driver implementation.
+type MetaStore interface {
+	// Stat returns the snapshot stat Info directly from
+	// the metadata.
+	Stat(ctx context.Context, key string) (snapshot.Info, error)
+
+	// Walk iterates through all metadata for the stored
+	// snapshots and calls the provided function for each.
+	Walk(ctx context.Context, fn func(context.Context, snapshot.Info) error) error
+
+	// CreateActive creates a new active snapshot transaction referenced by
+	// the provided key. The provided opts carries the parent and readonly
+	// properties to store with the active snapshots. Additionally a create
+	// method can be provided in order to allow the caller perform an action
+	// with the active snapshots identifier before the metadata is
+	// committed. This callback action should be as short as possible to
+	// avoid blocking the transaction longer than necessary.
+	CreateActive(ctx context.Context, key string, opts CreateActiveOpts) error
+
+	// GetActive returns the metadata for the active snapshot transaction
+	// referenced by the given key.
+	GetActive(ctx context.Context, key string, allParents bool) (Active, error)
+
+	// Remove removes a snapshot from the metastore. Cleanup is called
+	// during the transaction to allow the driver to make data unavailable
+	// before committing. Cleanup should run as fast as possible, and if
```

Comment:
Since this is done in a serialized transaction, other operations would be blocked by this. This is intended to allow for making the commit 2 phase (db and fs), the callback should then do the minimal amount to make the object unavailable but defer cleanup. Can also make this clearer in the documentation.